File: hornerm.cpp
#include "hornerm.h"

void hornerm(Matrix<double>& A, Matrix<double>& dif, Matrix<double>& Qc, Matrix<double>& fh, Matrix<double>& Qurmul)
{
	int N = size(dif, 1);
	int K = size(dif, 2);
	int n = size(Qc, 1); 

	for (int k = 1; k <= K; k++)
		for (int i = 1; i <= N; i++)
			Qurmul(i, k) = Qc(n)*dif(i, k);

	for (int k = 1; k <= K; k++)
	{
		for (int j = n - 1; j >= (1); j-=1)
		{
			if (Qc(j) != 0)
				for (int i = 1; i <= N; i++) fh(i, k) = Qc(j)*dif(i, k);
			else
				for (int i = 1; i <= N; i++) fh(i, k)=0;
			
			/** p=p+A*pAx = a*x+A*pAx **/
			if (global::int_form != 5) {
				matmult(A, Qurmul, fh);
			} else {
				matmult_dec(A, Qurmul, fh);
			}
			/** pAx= p = a*x+A*pAx **/
			for (int i = 1; i <= N; i++) Qurmul(i, k) = fh(i, k);
		}
	}
}

File: hornerm.m
function pAx=hornerm(A, x, a, p, pAx)

%int: k, n, N, i, j, int_form, K
%vector: a
%matrix: A, pAx, x, p 

global int_form;

N=length(x);

K=size(x, 2);
n=length(a); 

for k=1:K
  for i=1:N
    pAx(i, k)=a(n)*x(i, k);
  end
end

for k=1:K
  for j=n-1:-1:1
    % p=a*x
    if ( a(j) ~= 0 )
      for i = 1:N
	p(i, k)=a(j)*x(i, k);
      end
    else
      for i = 1:N
	p(i, k)=0;
      end
    end
    
    %p=p+A*pAx = a*x+A*pAx
    if ( int_form ~= 5 )
      p=matmult(A, pAx, p);
    else
      p=matmult_dec(A, pAx, p);
    end
    %pAx= p = a*x+A*pAx
    for i = 1:N
      pAx(i, k)=p(i, k);
    end
  end
end